addons/point_of_sale/models/pos_session.py

```diff
@@ -300,15 +309,534 @@ def _validate_session(self):
+            statement_cash_receivable_lines = cash_statement_lines.mapped('journal_entry_ids').filtered(lambda aml: aml.account_id.internal_type == 'receivable')
+            all_lines = (receivable_pos_cash_lines | statement_cash_receivable_lines)
+            accounts = all_lines.mapped('account_id')
+            lines_by_account = [all_lines.filtered(lambda l: l.account_id == account) for account in accounts]
+            for lines in lines_by_account:
+                lines.reconcile()
+
+    def reconcile_stock_output_lines(self):
+        """ Reconcile output lines in move_id of this session to the output lines of the account_move of stock_moves.
+        """
+        invoiced_account_moves = self.order_ids.filtered(lambda order: order.is_invoiced).mapped('account_move')
+        stock_moves = self.env['stock.move'].search([('picking_id', 'in', self.order_ids.picking_id.ids)])
+        account_moves = self.env['account.move'].search([('stock_move_id', 'in', stock_moves.ids)])
+        for out_account in stock_moves.mapped(lambda move: move.product_id.categ_id.property_stock_account_output_categ_id):
+            (account_moves | self.move_id | invoiced_account_moves)\
+                .mapped('line_ids')\
+                .filtered(lambda line: line.account_id.id == out_account.id).reconcile()
+
+    def _create_account_move_lines(self):
+        invoiced_orders = self.order_ids.filtered(lambda order: order.is_invoiced)
+        uninvoiced_orders = self.order_ids - invoiced_orders
+        lines_vals = []
+        lines_vals.extend(self._get_pos_receivable_lines_vals())
+        lines_vals.extend(self._get_invoiced_receivable_lines_vals(invoiced_orders))
+        lines_vals.extend(self._get_sales_lines_vals(uninvoiced_orders.mapped('lines')))
+        lines_vals.extend(self._get_tax_lines_vals(uninvoiced_orders.mapped('lines')))
+        lines_vals.extend(self._get_anglo_saxon_lines_vals(uninvoiced_orders))
+        lines_vals.extend(self._get_rounding_error_line_vals(lines_vals))
+        return self.env['account.move.line'].create(lines_vals)
+
+    def _validate_cash_statement(self, statement, statement_payments):
+        """ Validates each statement of the given session then returns
+            cash account.bank.statement.line records.
+        """
+        if not statement_payments:
+            return self.env['account.bank.statement.line']
+        cash_statement_lines = self._create_cash_statement_lines(statement, statement_payments)
+        statement_line_ids = cash_statement_lines.ids
+        if not self.config_id.cash_control:
+            statement.write({'balance_end_real': statement.balance_end})
+        statement.button_confirm_bank()
+        return self.env['account.bank.statement.line'].browse(statement_line_ids)
+
+    def _create_cash_statement_lines(self, statement, statement_payments):
```

Comment:
remove function _create_cash_statement_lines and create_vals and put it in _validate_cash_statement